src/StaticAnalysis/ExecutableLinesFindingVisitor.php

```diff
@@ -57,285 +30,325 @@
 final class ExecutableLinesFindingVisitor extends NodeVisitorAbstract
```

Comment:
I am opening another thread here to discuss the reasoning (result validity) only.

code:

```
    public function testBranchCoverage(): void
    {
        $a =
            mt_rand() === 0
            ?
            'a'
            .
            'a'
            :
            'b'
            .
            'b'
            ;

        try {fxA()
        .
        fxB()
        ;} catch (\Throwable $e) {}

        $b = true ? 'a' : fxA()
            .
            'c';

        $c = [


            'x'



        ];


        [


            'x'



        ];

        if (mt_rand() === 0) {
            var_dump($a . $b . count($c));
        }
    }
```

![image](https://user-images.githubusercontent.com/2228672/205281737-480dd619-8ed8-41aa-aebc-664470a9bb9b.png)

was whole marked as red, is it really possible it was executed? (is public and starting with test, other testXxx methods were executed/covered)

Reply:
Please elaborate further, I don't understand the environment nor you question

Reply:
I have added your PR branch as a dep, added the test code I posted to one phpunit test class...

It was completely red as there was not assertion (expected behaviour).

![image](https://user-images.githubusercontent.com/2228672/205288235-bed88927-7400-47ed-9888-4f4b6e0ccc56.png)

here are the objections I already posted

you must collect certain/maybe coverage points together with AST, and reason the maybe points based on AST node types/behaviour

Reply:
I don't know what's going on as it seems to me your are not using this PR correctly:

1. This PR doesn't mark your `L134` as executable
2. This PR does mark `L171`, `L174` and `L178` as executable and executed

I don't care about `L159`: if you put the `else` part of the ternary operator in a new line it become red again; users must use a proper coding style fixer to expect Line CC to be meaningful.

But, assuming `fxA()` is throwing an exception (which I don't know since you didn't provide the full code, but only a screenshot) I can confirm that `L154` must be green even though CC drivers are reporting it as not executed.
The same will go for method concatenation on objects:
```php
$myObject
    ->this_method_throws_an_exception()
    ->this_method_will_be_reported_as_executed_anyway()
;
```

The reason is that what you propose is impossible to be reliable, accurate and maintained.
For example:

```php
function foo($var) {
    return
        $var
        .
        'b'
        ;
}
function bar($var) {
    return
        $var
        .
        MyClass::FOO
        ;
}
```

Should we consider `$var` line as executable or not? The answer is it depends if `$var` is an object or not, and you can't know it only by parsing the code.
Let's compare current behaviour with this PR:

| Scenario | `9.2` report | PR report |
| --- | --- | --- |
| `foo` not called | :red_circle: `'b'` line red | :red_circle: all lines red |
| `foo` called smoothly | :heavy_check_mark: `'b'` line green | :heavy_check_mark: all lines green |
| `foo` called, `$var` object typecast raises an Exception | :red_circle: `'b'` line red | :heavy_check_mark: all lines green |
| `bar` not called | :red_circle: `MyClass::FOO` line red | :red_circle: all lines red |
| `bar` called smoothly | :heavy_check_mark: `MyClass::FOO` line green | :heavy_check_mark: all lines green |
| `bar` called, `$var` object typecast raises an Exception | :heavy_check_mark:/:red_circle: `MyClass::FOO` reported either red or green, depending on autoload runtimes | :heavy_check_mark: all lines green |

This PR is consistent, and where it fails to be precise on `'b'` and  `MyClass::FOO` lines you can enable Mutation Testing to check your tests are good enough.

Reply:
> I don't know what's going on as it seems to me your are not using this PR correctly:

repro code is here: https://github.com/atk4/ui/commit/5b9438914589de9feb22a22e1f74382a07572909

`L159` no, bad formatting must never even imply wrong (green or even red) coverage, If you will listen to me, and collect certain/maybe executable points together with AST node types, you can apply behavioural reasoning and achive valid output, this is also how phpstan and other tools handle AST

`fxA()` throws undefined function - https://3v4l.org/LlE6W

```
$myObject
    ->this_method_throws_an_exception()
    ->this_method_will_be_reported_as_executed_anyway()
    ->this_method_will_be_reported_as_executed_anyway()
;
```

again, this is a big problem, coverage must not lie

> Should we consider $var line as executable or not? The answer is it depends if $var is an object or not, and you can't know it only by parsing the code.

again, you cannot simply work with topmost AST node start-end range, you must collect executable points

example:

```
    return
        $var
        .
        'b'
        ;
```

`$var` and `'b'` should be analysed as maybe coverage points in an unconditional expr AST node and if the following statement (or an end of method/file or owning expr/branch) is reached (by xdebug data), only then you can reason these maybe coverage points (as covered/uncovered)

> table

(untested) I belive the problem with var is only when `$var` is const expr, a func param is never known upfront, thus `$var` coverage is reported by xdebug (but please, focus on the aka `L159` problem, not these unimportant details)

> This PR is consistent, and where it fails to be precise on 'b' and MyClass::FOO lines you can enable Mutation Testing to check your tests are good enough.

but it introduces many other problems and invalid output data

Reply:
> (but please, focus on the aka L159 problem, not these unimportant details)

> but it introduces many other problems and invalid output data

At this point we can only conclude we have different priorities: you consider more important a sometimes 100% correct report, while I consider more important an always 99% correct one.

I don't think this conversation can reach any better conclusion; the maintainers will choose what they prefer.

Have a nice day.

Reply:
> again, you cannot simply work with topmost AST node start-end range, you must collect executable points

BTW, this is exactly what we can't do.
For non executed files, this library has no exec points at all, but it must produce a report identical to what it would be with an execution

Reply:
I do not want to be manipulated at the first place and want to keep focus only on relevant things.

Yes, I expect 100% correct report. We shared this goal until you created this PR and the way I validated the output is by generating reports from more than 10k+ files and comparing the data againts real xdebug output at different PHP versions and PHP configurations. This output is asserted by the latest/`9.2.19` tests.

Saying "sometimes 100% correct report" is wrong and misleading. The report is always 100% correct. In quite small edge cases - compiled constant expressions (/w opcache enabled) at compile time are not present as not present in source/xdebug data. This can be addressed by proper AST branch analysis, but such improvement must never ever imply wrong coverage data.

Have a nice day too and I would be happy if we can agree on this and instead of replacing the correct actual/`9.2.19` executable analysis introduce another analyser, which will take these valid/stable/always correct data and compute optional branch analysis (for mutation testing) based on them. Everyone will be happy :)

Reply:
> In quite small edge cases - compiled constant expressions (/w opcache enabled) at compile time are not present as not present in source/xdebug data. This can be addressed by proper AST branch analysis

You can't solve #892 and foo/bar case from https://github.com/sebastianbergmann/php-code-coverage/pull/964#discussion_r1038268222 with AST analysis only.

If you can, code it to prove it please

> instead of replacing the correct actual/9.2.19 executable analysis

Except that it isn't correct because structurally flawed as proven in https://github.com/sebastianbergmann/php-code-coverage/pull/964#discussion_r1038268222 and this it is only sometimes 100% correct

Reply:
> You can't solve https://github.com/sebastianbergmann/php-code-coverage/pull/892 and foo/bar case from https://github.com/sebastianbergmann/php-code-coverage/pull/964#discussion_r1038268222 with AST analysis only.

for

```
<?php

function u() {
    return (string) mt_rand();
}
function x() {
    return // line 7
    u().
    1;
} // line 10
```

xdebug produces (uncached):

```
Array
(
    [4] => 1
    [5] => -2
    [8] => 1
    [9] => 1
    [10] => -2
)
```

and (cached):
```
Array
(
    [4] => 1
    [8] => 1
    [9] => 1
)
```

for `u` function declaration commented out (`u` function undefined) xdebug produces:

```
Array
(
    [8] => 1
    [9] => -1
)
```

for

```
<?php

function x() {
    return
    '1'.
    1; // line 6
}
```

it produces:

```
Array
(
    [6] => 1
)
```

so I would say it can be 100% correctly reasoned, as if an expression in a statement is const expr, the last expr line is (thankfully) always emit :)

the same for:
```
<?php

function foo($var) {
    return
        $var
        .
        'b' // line 7
        ;
}
```

xdebug produces:

```
Array
(
    [7] => 1
)
```

for (`$var` swapped/last):

```
<?php

function foo($var) {
    return
        'b'
        .
        $var
        ;
}
```

xdebug outputs (thankfully) the same/last expr line:

```
Array
(
    [7] => 1
)
```

Reply:
> the same for:
> 
> ```php
> function foo($var) {
>     return
>         $var
>         .
>         'b' // line 7
>         ;
> }
> ```
> 
> xdebug produces:
> 
> ```
> Array
> (
>     [7] => 1
> )
> ```

Exactly, and that's WRONG!

if `$var` is an object, line 5 becomes an executable line too, but you can't know it with a raw CC collection over an unexecuted function, so processing uncovered files would produce a different report from executed files, and that's unreliable

Reply:
🙏 understand `$var` variable fetch expression is not an executable line by php-src, thus it is not reported, this is also how other software report coverage lines...

I did an experiment to see if line 5 can become an executable line.

for:

```
<?php

function foo($var) {
    return
        $var
        .
        'b' // line 7
        ;
}

foo(new class () { public function __toString() { return (string) mt_rand(); } });
```

xdebug produces:

```
Array
(
    [7] => 1
    [11] => 1
)
```

with `opcache.enable_cli=0` I get:

```
Array
(
    [7] => 1
    [9] => -2
    [11] => 1
)
```

I am not strictly againts adding a variable fetch expression as an executable line if reasoned correctly, but I am not sure if mutation testing requires it as native xdebug never reported such line (at least not on my side, not "unstable"). I belive the reason is php-src/xdebug reports executable line for binary op, not for the elementary expressions.

In my previous post (and this too), I prooved the topmost expression in a statement does produce at least 1 coverage line. This is enought to reason all executable lines even the ones that do not produce any coverage.